src/connectors/s3.rs

```diff
@@ -0,0 +1,373 @@
+        let (client_sender, client_receiver) = channel::bounded(2);
+
+        // set the sender and receiver for a new client
+        self.client_sender = Some(client_sender);
+
+        let s3_sink = S3Sink {
+            client_receiver,
+            bucket: self.config.bucket.clone(),
+            client: None,
+            buffer: Vec::new(),
+            current_key: "".to_owned(),
+            parts: Vec::new(),
+            upload_id: "".to_owned(),
+            part_number: 0,
+            min_part_size: self.config.min_part_size,
+        };
+
+        let addr = builder.spawn(s3_sink, sink_context)?;
+        Ok(Some(addr))
+    }
+
+    // Runs on need to connect/reconnect. (multiple times)
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        // Create the client and establish the connection.
+        let s3_config = s3::config::Config::builder()
+            .credentials_provider(Credentials::from_keys(
+                self.config.aws_access_token.clone(),
+                self.config.aws_secret_access_key.clone(),
+                None,
+            ))
+            .region(Region::new(self.config.aws_region.clone()));
+
+        let s3_config = match &self.endpoint {
+            Some(uri) => s3_config.endpoint_resolver(Endpoint::immutable(uri.clone())), // unwrap checked before in config generation
+            None => (s3_config),
+        };
+
+        let client = S3Client::from_conf(s3_config.build());
+
+        // Check for the existence of the bucket.
+        let exist_bucket = client
+            .head_bucket()
+            .bucket(self.config.bucket.clone())
+            .send()
+            .await;
+
+        if exist_bucket.is_err() {
+            return Err(exist_bucket.unwrap_err().into());
+        }
+
+        // Send the client to the sink.
+        self.client_sender.as_ref().unwrap().send(client).await?;
+
+        Ok(true)
+    }
+
+    // FIXME: text or json
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+}
+
+impl S3Sink {
+    async fn set_new_key(&mut self, key: String) -> Result<()> {
+        // Finish the previous multipart upload if any.
+        if self.current_key != "" {
+            self.complete_multipart().await?;
+        }
+
+        if key != "" {
+            self.initiate_multipart(key).await?;
+        }
+        // NOTE: The buffers are cleared when the stuff is commited.
+        Ok(())
+    }
+
+    async fn initiate_multipart(&mut self, key: String) -> Result<()> {
+        self.current_key = key;
+        self.part_number = 0; // Reset to new sequence.
+        self.upload_id = self
+            .client
+            .as_ref()
+            .unwrap()
+            .create_multipart_upload()
+            .bucket(self.bucket.clone())
+            .key(self.current_key.clone())
+            .send()
+            .await
+            .map(|resp| resp.upload_id.unwrap())?; // This unwrap is safe as we have a successful response.
```

Comment:
could you make this an `and_then` and turn the `upload_id` option into a result with `ok_or_else`?